src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -1867,5 +1868,2119 @@ void verify(PropertyPatternClauseSyntax clause, string syntax, string type)
+                // (16,27): error CS0656: Missing compiler required member 'System.Index..ctor'
+                //         _ = new C<int?>()[^1]; // 6, 7
+                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "^1").WithArguments("System.Index", ".ctor").WithLocation(16, 27),
+                // (18,33): error CS9202: Length patterns may not be used for a value of type 'C<string?>'.
+                //         if (new C<string?>() is [var length3]) // 8
+                Diagnostic(ErrorCode.ERR_UnsupportedTypeForLengthPattern, "[var length3]").WithArguments("C<string?>").WithLocation(18, 33),
+                // (20,30): error CS0518: Predefined type 'System.Index' is not defined or imported
+                //         _ = new C<string?>()[^1]; // 9, 10
+                Diagnostic(ErrorCode.ERR_PredefinedTypeNotFound, "^1").WithArguments("System.Index").WithLocation(20, 30),
+                // (20,30): error CS0656: Missing compiler required member 'System.Index..ctor'
+                //         _ = new C<string?>()[^1]; // 9, 10
+                Diagnostic(ErrorCode.ERR_MissingPredefinedMember, "^1").WithArguments("System.Index", ".ctor").WithLocation(20, 30)
+                );
+
+            var tree = compilation.SyntaxTrees.Single();
+            var model = compilation.GetSemanticModel(tree, ignoreAccessibility: false);
+            var declarations = tree.GetRoot().DescendantNodes().OfType<VarPatternSyntax>().ToArray();
+
+            // Note the type is always int
+            verify(declarations[0], "length", "System.Int32");
+            verify(declarations[1], "length2", "System.Int32");
+            verify(declarations[2], "length3", "System.Int32");
+
+            void verify(VarPatternSyntax declaration, string name, string expectedType)
+            {
+                var local = (ILocalSymbol)model.GetDeclaredSymbol(declaration.Designation)!;
+                Assert.Equal(name, local.Name);
+                Assert.Equal(expectedType, local.Type.ToTestDisplayString(includeNonNullable: true));
+            }
+        }
+
+        [Fact]
+        public void ListPattern_GenericIndexingParameter()
+        {
+            var source = @"
+#nullable enable
+class C<T>
+{
+    public int Length => throw null!;
+    public int this[T i] => throw null!;
+
+    public void M()
+    {
+        if (new C<int>() is { var item }) // 1
+            item.ToString();
+        _ = new C<int>()[^1]; // 2
+
+        if (new C<int?>() is { var item2 }) // 3
+            item2.Value.ToString();
+        _ = new C<int?>()[^1]; // 4
+
+        if (new C<System.Index>() is { var item3 })
+            item3.ToString();
+        _ = new C<System.Index>()[^1];
+
+        if (new C<System.Index?>() is { var item4 })
+            item4.ToString();
+        _ = new C<System.Index?>()[^1];
+    }
+}
+";
+            // PROTOTYPE errors 1 and 2 are unexpected
```

Comment:
This is probably because we expect exactly `this[int]` as the member (vs. only a viable lookup for Index).
We're checking OriginalDefinition, which I think is per spec.